aztk/utils.py

```diff
@@ -0,0 +1,210 @@
+import getpass
+import time
+from typing import List
+from subprocess import call
+import azure.batch.models as batch_models
+import aztk_sdk.spark
+import aztk_sdk.utils.error as error
+import aztk_sdk.utils.get_ssh_key as get_ssh_key
+from . import log
+
+
+def get_ssh_key_or_prompt(ssh_key, username, password, secrets_config):
+    ssh_key = get_ssh_key.get_user_public_key(ssh_key, secrets_config)
+
+    if username is not None and password is None and ssh_key is None:
+        password = getpass.getpass("Please input a password for user '{0}': ".format(username))
+        confirm_password = getpass.getpass("Please confirm your password for user '{0}': ".format(username))
+        if password != confirm_password:
+            raise error.AztkError("Password confirmation did not match, please try again.")
+        if not password:
+            raise error.AztkError(
+                "Password is empty, cannot add user to cluster. Provide a ssh public key in .aztk/secrets.yaml. Or provide an ssh-key or password with commnad line parameters (--ssh-key or --password).")
+    return ssh_key, password
+
+def print_cluster(cluster: aztk_sdk.spark.models.Cluster, batch_client):
+    node_count = __pretty_node_count(cluster)
+
+    log.info("")
+    log.info("Cluster         %s", cluster.id)
+    log.info("------------------------------------------")
+    log.info("State:          %s", cluster.visible_state)
+    log.info("Node Size:      %s", cluster.vm_size)
+    log.info("Nodes:          %s", node_count)
+    log.info("| Dedicated:    %s", __pretty_dedicated_node_count(cluster))
+    log.info("| Low priority: %s", __pretty_low_pri_node_count(cluster))
+    log.info("")
+
+    print_format = '{:<36}| {:<15} | {:<21}| {:<8}'
+    print_format_underline = '{:-<36}|{:-<17}|{:-<22}|{:-<8}'
+    log.info(print_format.format("Nodes", "State", "IP:Port", "Master"))
+    log.info(print_format_underline.format('', '', '', ''))
+
+    if not cluster.nodes:
+        return
+    for node in cluster.nodes:
+        ip, port = __get_connection_info(cluster.id, node.id, batch_client)
+        log.info(print_format.format(node.id, node.state.value, '{}:{}'.format(ip, port),
+                                        '*' if node.id == cluster.master_node_id else ''))
+    log.info('')
+
+def __get_connection_info(pool_id, node_id, batch_client):
+    """
+    Get connection info of specified node in pool
+    :param batch_client: The batch client to use.
+    :type batch_client: `batchserviceclient.BatchServiceClient`
+    :param str pool_id: The pool id to look up
+    :param str node_id: The node id to look up
+    """
+    rls = batch_client.compute_node.get_remote_login_settings(
```

Comment:
Is utils supposed to be part of the sdk or the cli? If part of the cli, then I think it's weird that we are using a batch_client call here since the cli should ideally never need to know about it.

Reply:
same goes for all batch.foo calls in this file

Reply:
you are right -- the separation here is not clear enough. All functions that rely on the batch_client should be moved to the sdk.

Basically __get_connection_info should be moved to the sdk somewhere.

I'll have to think on the direct calls in ssh_in_master() because I feel like SSHing is a pure CLI experience.